Failed to build from source, aarch64, got /usr/bin/ld: ../src/.libs/libunwind.so: undefined reference to `__aarch64_cas8_acq_rel'

Full compilation log: https://download.copr.fedorainfracloud.org/results/packit/skupperproject-skupper-router-1014/fedora-rawhide-aarch64/06847035-skupper-router/builder-live.log.gz

```
Making install in tests
make[1]: Entering directory '/builddir/build/BUILD/libunwind-1.8.0-rc1/tests'
  CC       forker.o
  CCLD     forker
  CC       Gperf-simple.o
  CCLD     Gperf-simple
/usr/bin/ld: ../src/.libs/libunwind.so: undefined reference to `__aarch64_cas8_acq_rel'
/usr/bin/ld: ../src/.libs/libunwind.so: undefined reference to `__aarch64_swp1_acq_rel'
/usr/bin/ld: ../src/.libs/libunwind.so: undefined reference to `__aarch64_ldadd8_acq_rel'
/usr/bin/ld: ../src/.libs/libunwind.so: undefined reference to `__aarch64_ldadd4_acq_rel'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:1204: Gperf-simple] Error 1
make[1]: Leaving directory '/builddir/build/BUILD/libunwind-1.8.0-rc1/tests'
make: *** [Makefile:618: install-recursive] Error 1
```

Tried to do

```
+ ./configure --prefix=/builddir/build/BUILD/libunwind-install CFLAGS=-latomic LDFLAGS=-latomic
```

and then I got

```
In file included from aarch64/Los-linux.c:4:
aarch64/Gos-linux.c: In function â€˜_ULaarch64_local_resumeâ€™:
aarch64/Gos-linux.c:41:7: error: â€˜asmâ€™ operand has impossible constraints
   41 |       __asm__ __volatile__ (
      |       ^~~~~~~
make[2]: *** [Makefile:4381: aarch64/Los-linux.lo] Error 1
make[2]: Leaving directory '/builddir/build/BUILD/libunwind-1.8.0-rc1/src'
make[1]: *** [Makefile:4527: install] Error 2
make[1]: Leaving directory '/builddir/build/BUILD/libunwind-1.8.0-rc1/src'
make: *** [Makefile:618: install-recursive] Error 1
error: Bad exit status from /var/tmp/rpm-tmp.UIEGPW (%build)
```

full log at https://download.copr.fedorainfracloud.org/results/packit/skupperproject-skupper-router-1014/fedora-39-aarch64/06847997-skupper-router/builder-live.log.gz

I have a similar issue on openSUSE Tumbleweed aarch64:
```
[   79s] libtool: link: gcc -fno-optimize-sibling-calls -mbranch-protection=standard -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__ -flto=auto -o .libs/Lperf-trace Lperf-trace.o  ../src/.libs/libunwind.so -lc -llzma -Wl,-rpath -Wl,/usr/lib64
[   79s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: .libs/Lperf-trace: hidden symbol `__aarch64_cas8_acq_rel' in /usr/lib64/gcc/aarch64-suse-linux/13/libgcc.a(cas_8_4.o) is referenced by DSO
[   79s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: final link failed: bad value
[   79s] collect2: error: ld returned 1 exit status
[   79s] make[1]: *** [Makefile:1323: Lperf-trace] Error 1
```

> I have a similar issue on openSUSE Tumbleweed aarch64:
> 
> ```
> [   79s] libtool: link: gcc -fno-optimize-sibling-calls -mbranch-protection=standard -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__ -flto=auto -o .libs/Lperf-trace Lperf-trace.o  ../src/.libs/libunwind.so -lc -llzma -Wl,-rpath -Wl,/usr/lib64
> [   79s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: .libs/Lperf-trace: hidden symbol `__aarch64_cas8_acq_rel' in /usr/lib64/gcc/aarch64-suse-linux/13/libgcc.a(cas_8_4.o) is referenced by DSO
> [   79s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: final link failed: bad value
> [   79s] collect2: error: ld returned 1 exit status
> [   79s] make[1]: *** [Makefile:1323: Lperf-trace] Error 1
> ```

As a workaround, we can disable LTO.

> @jiridanek @ggardet please check to see if #732 resolves the issue of failing to find the atomic symbols for you.

For openSUSE Tumbleweed with #732 and LTO disbaled, it still fails:
```
[  137s] checking if -latomic is required... no
```
```
[  159s] /usr/bin/bash ../libtool  --tag=CC   --mode=link gcc -fno-optimize-sibling-calls -mbranch-protection=standard -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__  -flto=auto -o Lperf-trace Lperf-trace.o ../src/libunwind.la 
[  159s] libtool: link: gcc -fno-optimize-sibling-calls -mbranch-protection=standard -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -fexceptions -Wall -Wsign-compare -D__EXTENSIONS__ -flto=auto -o .libs/Lperf-trace Lperf-trace.o  ../src/.libs/libunwind.so -lc -llzma -Wl,-rpath -Wl,/usr/lib64
[  159s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: .libs/Lperf-trace: hidden symbol `__aarch64_cas8_acq_rel' in /usr/lib64/gcc/aarch64-suse-linux/13/libgcc.a(cas_8_4.o) is referenced by DSO
[  159s] /usr/lib64/gcc/aarch64-suse-linux/13/../../../../aarch64-suse-linux/bin/ld: final link failed: bad value
[  159s] collect2: error: ld returned 1 exit status
```

--- a/configure.ac
+++ b/configure.ac
@@ -81,17 +81,6 @@ AC_CHECK_TYPES([struct elf_prstatus, struct prstatus, procfs_status, elf_fpregse
 
 dnl Checks for libraries.
 AC_MSG_NOTICE([--- Checking for libraries ---])
-save_LDFLAGS="$LDFLAGS"
-save_LIBS="$LIBS"
-LDFLAGS="${LDFLAGS} -nostdlib"
-AC_SEARCH_LIBS([_Unwind_Resume], [gcc_s gcc],
-               [AS_IF([test "$ac_cv_search__Unwind_Resume" != "none required"],
-                      [AC_SUBST([LIBCRTS], ["$ac_cv_search__Unwind_Resume"])])],
-               [],
-               [-lc]
-)
-LIBS="$save_LIBS"
-LDFLAGS="$save_LDFLAGS"
 AC_SEARCH_LIBS([__uc_get_grs], [uca])
 AC_SEARCH_LIBS([getcontext], [ucontext])
 
@@ -299,7 +288,7 @@ AC_ARG_ENABLE([debug],
 )
 AC_MSG_RESULT([$enable_debug])
 
-AC_MSG_CHECKING([if C++ exception support shouldm  be built])
+AC_MSG_CHECKING([if C++ exception support should be built])
 AC_ARG_ENABLE([cxx_exceptions],
               [AS_HELP_STRING([--enable-cxx-exceptions],
                               [use libunwind to handle C++ exceptions
@@ -314,6 +303,16 @@ AS_IF([test $enable_cxx_exceptions = check],
 )
 AC_MSG_RESULT([$enable_cxx_exceptions])
 AM_CONDITIONAL([SUPPORT_CXX_EXCEPTIONS], [test x$enable_cxx_exceptions = xyes])
+AM_COND_IF([SUPPORT_CXX_EXCEPTIONS],
+           [save_LDFLAGS="$LDFLAGS"
+            LDFLAGS="${LDFLAGS} -nostdlib"
+            AC_SEARCH_LIBS([_Unwind_Resume], [gcc_s gcc],
+                           [AS_IF([test "$ac_cv_search__Unwind_Resume" != "none required"],
+                                  [AC_SUBST([UNW_CRT_LIBADD],  ["-lc $ac_cv_search__Unwind_Resume"])
+                                   AC_SUBST([UNW_CRT_LDFLAGS], ["-WCClinker -nostdlib"])])]
+                           [-lc])
+            LDFLAGS="$save_LDFLAGS"]
+)
 
 AC_MSG_CHECKING([if documentation should be built])
 AC_ARG_ENABLE([documentation],

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -127,7 +127,7 @@ endif
 # If local unwinding is being built, link in the local unwinding functions
 libunwind_libadd =
 if !REMOTE_ONLY
-  libunwind_libadd += libunwind.la -lc
+  libunwind_libadd += libunwind.la
 endif
 
 ### pkg-config:
@@ -1264,10 +1264,13 @@ endif
 # Don't link with standard libraries, because those may mention
 # libunwind already.
 #
-libunwind_la_LDFLAGS =	$(COMMON_SO_LDFLAGS) -XCClinker -nostdlib \
-			$(LDFLAGS_STATIC_LIBCXA) -version-info $(SOVERSION)
-libunwind_la_LIBADD  += -lc $(LIBCRTS)
-libunwind_la_LIBADD += $(LIBLZMA) $(LIBZ)
+libunwind_la_LDFLAGS = $(COMMON_SO_LDFLAGS) \
+                       $(UNW_CRT_LDFLAGS) \
+                       $(LDFLAGS_STATIC_LIBCXA) \
+                       -version-info $(SOVERSION)
+libunwind_la_LIBADD += $(UNW_CRT_LIBADD) \
+                       $(LIBLZMA) \
+                       $(LIBZ)
 
 noinst_HEADERS += unwind/unwind-internal.h
 

